### ISSUE TITLE
FileSystem.readAsStringAsync not readable in iOS

### ISSUE THREAD

**Issue by zhouhao27** [expo/expo#31310]
### Minimal reproducible example

https://github.com/zhouhao27/readAsStringAsync-Bug

### What platform(s) does this occur on?

iOS

### Where did you reproduce the issue?

in a development build

### Summary

- Created app with `npx create-expo-app@latest` 
- Add `expo-image-picker` and `expo-file-system`.
- To start the application in iOS: `npx expo run:ios`
- Then in home page, click the image picker to pick up an image at the same time to read the file. But it shows the following error message:

> [Error: Calling the 'readAsStringAsync' function has failed
→ Caused by: File '/Users/username/Library/Developer/CoreSimulator/Devices/162C900B-C120-4DD7-A7CF-A12C11599575/data/Containers/Data/Application/F3B43C4A-FE82-40D5-9B42-6EFEB5691661/Library/Caches/ImagePicker/DEF6A55B-546A-43DB-8280-0A169AF83944.jpg' is not readable]

But I goto the file path, the jpg file is there and is readable.

### Environment

expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 15.0
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 20.15.0 - ~/.nvm/versions/node/v20.15.0/bin/node
      Yarn: 1.22.22 - ~/.nvm/versions/node/v20.15.0/bin/yarn
      npm: 10.7.0 - ~/.nvm/versions/node/v20.15.0/bin/npm
      Watchman: 2024.06.17.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.15.2 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 24.0, iOS 18.0, macOS 15.0, tvOS 18.0, visionOS 2.0, watchOS 11.0
    IDEs:
      Android Studio: 2024.1 AI-241.18034.62.2411.12169540
      Xcode: 16.0/16A5171r - /usr/bin/xcodebuild
    npmPackages:
      expo: ~51.0.28 => 51.0.31
      expo-router: ~3.5.23 => 3.5.23
      react: 18.2.0 => 18.2.0
      react-dom: 18.2.0 => 18.2.0
      react-native: 0.74.5 => 0.74.5
      react-native-web: ~0.19.10 => 0.19.12
    Expo Workflow: bare

### Expo Doctor Diagnostics

✔ Check Expo config for common issues
✔ Check package.json for common issues
✔ Check dependencies for packages that should not be installed directly
✔ Check for common project setup issues
✔ Check for issues with metro config
✔ Check npm/ yarn versions
✔ Check if the project meets version requirements for submission to app stores
✔ Check Expo config (app.json/ app.config.js) schema
✔ Check for app config fields that may not be synced in a non-CNG project
✔ Check native tooling versions
✔ Check that native modules do not use incompatible support packages
✔ Check for legacy global CLI installed locally
✔ Check that packages match versions required by installed Expo SDK
✔ Check that native modules use compatible support package versions for installed Expo SDK

**Comment by douglasjunior** [comment]
Samer here, using `expo-image-picker`, `expo-document-picker` or `react-native-image-crop-picker`  the `FileSystem.readAsStringAsync` always return:

```
[Error: Calling the 'readAsStringAsync' function has failed
→ Caused by: File '/var/mobile/Containers/Data/Application/2585E2A4-F214-41D0-8A26-AE2134EC442C/Library/Caches/ImagePicker/5D562492-EFAA-4AA4-945F-A78B2E7EEC59.png' is not readable]
```

```
[Error: Calling the 'readAsStringAsync' function has failed
→ Caused by: File '/private/var/mobile/Containers/Data/Application/2585E2A4-F214-41D0-8A26-AE2134EC442C/tmp/react-native-image-crop-picker/3A766AC3-30E9-40BF-AB3C-439D4EFFF4EA.jpg' is not readable]
```

```
[Error: Calling the 'readAsStringAsync' function has failed
→ Caused by: File '/var/mobile/Containers/Data/Application/2585E2A4-F214-41D0-8A26-AE2134EC442C/Library/Caches/DocumentPicker/2A570B91-3C6E-43CC-9D89-CE452B04283F.pdf' is not readable]
```

Environment:

```
  expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.6.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 18.17.1 - ~/.nvm/versions/node/v18.17.1/bin/node
      Yarn: 3.6.4 - ~/.nvm/versions/node/v18.17.1/bin/yarn
      npm: 9.6.7 - ~/.nvm/versions/node/v18.17.1/bin/npm
    Managers:
      CocoaPods: 1.14.2 - /Users/douglas.junior/.rbenv/shims/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.5, iOS 17.5, macOS 14.5, tvOS 17.5, visionOS 1.2, watchOS 10.5
      Android SDK:
        API Levels: 30, 33, 34
        Build Tools: 30.0.3, 33.0.0, 33.0.1, 33.0.2, 34.0.0
        System Images: android-34 | Google APIs ARM 64 v8a
    IDEs:
      Android Studio: 2024.1 AI-241.15989.150.2411.11948838
      Xcode: 15.4/15F31d - /usr/bin/xcodebuild
    npmPackages:
      expo: 51.0.32 => 51.0.32 
      react: 18.2.0 => 18.2.0 
      react-dom: 18.2.0 => 18.2.0 
      react-native: 0.74.3 => 0.74.3 
    Expo Workflow: bare
```

Versions:

```
    "expo": "51.0.32",
    "expo-application": "5.9.1",
    "expo-camera": "15.0.14",
    "expo-document-picker": "12.0.2",
    "expo-file-system": "17.0.1",
    "expo-image-picker": "15.0.7",
    "expo-linear-gradient": "13.0.2",
    "expo-secure-store": "13.0.2",
```

**Comment by douglasjunior** [comment]
Just changing `enconding` option to `base64` fixes to me. 😓

```diff
  const content = await ExpoFileSystem.readAsStringAsync(
    fileUri,
    {
-      encoding: 'utf8',
+      encoding: 'base64',
    },
  );
```

**Comment by aleqsio** [comment]
The native error being thrown here is this:

```Error Domain=NSCocoaErrorDomain Code=261 "The file “1EE61921-FAE1-40E6-86F9-223FF51BC957.jpg” couldn’t be opened using text encoding Unicode (UTF-8)."```

To be honest, I don't see what is the usecase for reading a `.jpg` file as a string – perhaps you'd prefer reading that as a blob or a typed array? We can add it to expo-file-system@next.

**Comment by douglasjunior** [comment]
I agree @aleqsio, but I have no control over this API and they request the documents in base64.

**Comment by paraita** [comment]
I'm also seeing this error when running a build (with a custom domain text file):

```
Calling the 'readAsStringAsync' function has failed
→ Caused by: File '/Users/paraita/Library/Developer/CoreSimulator/Devices/2F05A9D5-5FFC-406A-A75F-05C81DDCD5F4/data/Containers/Bundle/Application/E00B2B6C-35DD-4DE0-8E81-91B4AFA2C17E/Paraitapp1.app/assets/_common/machine/_assets/XYZ.jsonasset' is not readable
```

That XYZ.jsonasset file is indeed here and is correctly read when running in development mode.

**Comment by LukasB-DEV** [comment]
@paraita does not work for me. Still the same error with 
`const jsonString = await FileSystem.readAsStringAsync(filePath, {
    encoding: EncodingType.UTF8
});`

Any solution?

**Comment by LukasB-DEV** [comment]
> The native error being thrown here is this:
> 
> `Error Domain=NSCocoaErrorDomain Code=261 "The file “1EE61921-FAE1-40E6-86F9-223FF51BC957.jpg” couldn’t be opened using text encoding Unicode (UTF-8)."`
> 
> To be honest, I don't see what is the usecase for reading a `.jpg` file as a string – perhaps you'd prefer reading that as a blob or a typed array? We can add it to expo-file-system@next.

same issue for other files. I want to read a gpx file. There we need to read as string.

**Comment by LukasB-DEV** [comment]
@aleqsio any update?

**Comment by LukasB-DEV** [comment]
How can I use expo-file-system/next? According to the documentation, it is included in expo-file-system. However, I can't import it and it doesn't seem to be included (github).

**Comment by rubenupmeet** [comment]
I encountered the same issue:

Saving the file in UTF-8.
Trying to retrieve it by specifying UTF-8.
I simply had to remove the UTF-8 specification when retrieving it, and it worked!

**Comment by nemoneph** [comment]
@LukasB-DEV : same issue (also with gpx!) did you manage to resolve this issue ?
